L1Trigger/TrackerTFP/src/MiniHoughTransform.cc

```diff
@@ -0,0 +1,285 @@
+      const TTDTC::Stream& stream = streams[region_ * numBinsInv2R_ + binInv2R];
+      nStubsHT += accumulate(stream.begin(), stream.end(), 0, valid);
+    }
+    stubsHT_.reserve(nStubsHT);
+    stubsMHT_.reserve(nStubsHT * numCells_);
+    for (int binInv2R = 0; binInv2R < numBinsInv2R_; binInv2R++) {
+      const int inv2R = inv2R_.toSigned(binInv2R);
+      const TTDTC::Stream& stream = streams[region_ * numBinsInv2R_ + binInv2R];
+      vector<StubHT*>& stubs = input_[binInv2R];
+      stubs.reserve(stream.size());
+      for (const TTDTC::Frame& frame : stream) {
+        StubHT* stub = nullptr;
+        if (frame.first.isNonnull()) {
+          stubsHT_.emplace_back(frame, dataFormats_, inv2R);
+          stub = &stubsHT_.back();
+        }
+        stubs.push_back(stub);
+      }
+    }
+  }
+
+  void MiniHoughTransform::produce(TTDTC::Streams& accepted, TTDTC::Streams& lost) {
+    // fill MHT cells
+    vector<deque<StubMHT*>> stubsCells(numBinsInv2R_ * numCells_);
+    for (int channel = 0; channel < numBinsInv2R_; channel++)
+      fill(input_[channel], stubsCells, channel);
+    // perform static load balancing
+    vector<vector<StubMHT*>> streamsSLB(numBinsInv2R_);
+    for (int channel = 0; channel < numBinsInv2R_; channel++) {
+      vector<deque<StubMHT*>> tmp(numCells_);
+      for (int k = 0; k < numCells_; k++)
+        swap(tmp[k], stubsCells[(channel / numCells_) * numBinsInv2R_ + channel % numCells_ + k * numCells_]);
+      slb(tmp, streamsSLB[channel], lost[channel]);
+    }
+    // dynamic load balancing stage 1
+    vector<vector<StubMHT*>> streamsDLB(numBinsInv2R_);
+    for (int node = 0; node < numNodes_; node++) {
+      vector<vector<StubMHT*>> tmp(numChannel_);
+      for (int k = 0; k < numChannel_; k++)
+        swap(tmp[k], streamsSLB[(node / numCells_) * numNodes_ + node % numCells_ + k * numCells_]);
+      dlb(tmp);
+      for (int k = 0; k < numChannel_; k++)
+        swap(tmp[k], streamsDLB[node * numChannel_ + k]);
+    }
+    // dynamic load balancing stage 2
+    vector<vector<StubMHT*>> streamsMHT(numBinsInv2R_);
+    for (int node = 0; node < numNodes_; node++) {
+      vector<vector<StubMHT*>> tmp(numChannel_);
+      for (int k = 0; k < numChannel_; k++)
+        swap(tmp[k], streamsDLB[node + k * numNodes_]);
+      dlb(tmp);
+      for (int k = 0; k < numChannel_; k++)
+        swap(tmp[k], streamsMHT[node * numChannel_ + k]);
+    }
+    // fill output product
+    for (int channel = 0; channel < numBinsInv2R_; channel++) {
+      const vector<StubMHT*>& stubs = streamsMHT[channel];
+      TTDTC::Stream& stream = accepted[region_ * numBinsInv2R_ + channel];
+      stream.reserve(stubs.size());
+      for (StubMHT* stub : stubs)
+        stream.emplace_back(stub ? stub->frame() : TTDTC::Frame());
+    }
+  }
+
+  // perform finer pattern recognition per track
+  void MiniHoughTransform::fill(const vector<StubHT*>& stubs, vector<deque<StubMHT*>>& streams, int channel) {
```

Comment:
To make it easier to distinguish input from output args, please list input ones first. e.g. Make "channel" first arg instead of last one.

Reply:
done